### PR TITLE
feat: introduce scaffold admission-request command

### PR DESCRIPTION
**Pull request by flavio**
This command creates a Kubernetes AdmissionRequest object starting from a Kubernetes Resource definition.

The following command:

```console
kwctl scaffold \
  admission-request \
  --operation CREATE \
  --object ingress.yaml
```

produces an AdmissionRequest CREATE event for the given Ingress object.

Currently, only the CREATE operation is supported. The other operations will be added in the future.

### CODE REVIEW

**Comment by flavio** [comment]
Depends on https://github.com/kubewarden/policy-evaluator/pull/548

File: README.md [hunk 1]
```diff
@@ -299,16 +312,16 @@ Then start a new shell or run `source ~/.zshrc` once.
...
+When you download a [kwctl release](https://github.com/kubewarden/kwctl/releases/) each zip file contains two
 files that can be used for verification: `kwctl.sig` and `kwctl.pem`.
 
 In order to verify kwctl you need cosign installed, and then execute the following command:
 
 ```
 cosign verify-blob \
   --signature kwctl-linux-x86_64.sig \
-  --cert kwctl-linux-x86_64.pem kwctl-linux-x86_64 
+  --cert kwctl-linux-x86_64.pem kwctl-linux-x86_64
```
Comment (viccuad) [inline README.md:324]:
```suggestion
  --cert kwctl-linux-x86_64.pem kwctl-linux-x86_64 \
```

File: src/cli.rs [hunk 2]
```diff
@@ -486,6 +486,25 @@ fn subcommand_scaffold() -> Command {
...
+            .short('o')
+            .required(true)
+            .value_name("TYPE")
+            .value_parser(PossibleValuesParser::new(["CREATE"])) //TODO: add UPDATE and DELETE
+            .help("Kubewarden Custom Resource type"),
+        Arg::new("object")
+            .long("object")
+            .value_name("PATH")
+            .help("The file containing the new object being admitted"),
+        Arg::new("old-object")
```
Comment (viccuad) [inline src/cli.rs:501]: maybe worth commenting until we have UPDATE.

File: src/scaffold/admission_request.rs [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+    pub static ref DEFAULT_ROOT: ProjectDirs =
+        ProjectDirs::from("io.kubewarden", "", "kubewarden").unwrap();
+    pub static ref DEFAULT_KWCTL_CACHE: PathBuf = DEFAULT_ROOT.cache_dir().join("kwctl");
+    pub static ref RESOURCE_CATALOG_FILE: PathBuf =
+        DEFAULT_KWCTL_CACHE.join("resource_catalog.json");
+}
+
+const FALLBACK_API_RESOURCE_PLURAL_NAME: &str = "this-is-the-plural-name-of-the-resource-this-information-is-not-used-by-policies-and-requires-a-connection-to-an-api-server-to-be-obtained";
+
+/// Types of AdmissionRequest operation we can scaffolf
```
Comment (viccuad) [inline src/scaffold/admission_request.rs:35]:
```suggestion
/// Types of AdmissionRequest operation we can scaffold
```

File: src/scaffold/admission_request.rs [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+    object: Option<PathBuf>,
+    old_object: Option<PathBuf>,
+) -> Result<()> {
+    validate_params(&operation, object.as_ref(), old_object.as_ref())?;
+
+    let output = match operation {
+        Operation::Create => {
+            scaffold_create(
+                RESOURCE_CATALOG_FILE.to_path_buf(),
+                kube_client,
```
Comment (viccuad) [inline src/scaffold/admission_request.rs:262]: are we using the correct kube_client here? shouldn't we be using the newly created function?

**Review by viccuad — COMMENTED** [review]
LGTM, just one doubt about kube_client.

File: src/cli.rs (hunk 2 above)
Comment (flavio) [reply src/cli.rs:501]: I would like to do that with another PR, right after this one gets merged

File: src/scaffold/admission_request.rs (hunk 4 above)
Comment (flavio) [reply src/scaffold/admission_request.rs]: We were using the right function, the naming was not making that evident, though.

I've fixed that with ddd8916b938fe778e42e5c8c5712131ef116d28f

**Review by viccuad — APPROVED** [review]

**Review by jvanz — COMMENTED** [review]
LGTM. But I have a question in mind, what happens if the user has multiple contexts in the kubeconfig? Will the catalog contains the kinds from the default context? I'm wondering  about the UX for the user which is building a request for a CRDs which is not in the "default" context. 

If this is an issue at all,  I don't think this is a huge issue as well and we do not need to fix it now. But if it is, is there an easy workaround for it? Like defining the context to be used or fetch resources from all the contexts ( I'm not sure if this is a good idea in end end as well ) .

**Comment by flavio** [comment]
> LGTM. But I have a question in mind, what happens if the user has multiple contexts in the kubeconfig? Will the catalog contains the kinds from the default context? I'm wondering about the UX for the user which is building a request for a CRDs which is not in the "default" context.

The catalog will be built from the default context.

> If this is an issue at all, I don't think this is a huge issue as well and we do not need to fix it now. But if it is, is there an easy workaround for it? Like defining the context to be used or fetch resources from all the contexts ( I'm not sure if this is a good idea in end end as well ) .

Going through all the contexts would take time and also possible timeouts because maybe some of the clusters are not reachable.

It's good that you pointed out this possible limitation, but I would not focus on that right now. The scaffold code always generate something that will work for 99.9% of the policies.

We can find a solution to this corner case situation if someone complains about that.

**Review by jvanz — APPROVED** [review]

File: src/scaffold/admission_request.rs [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+    resources: HashMap<String, APIResource>,
+    #[serde(skip)]
+    restored_from: ApiResourceCatalogRestoredFrom,
+}
+
+impl ApiResourceCatalog {
+    /// Create a new catalog. The function will try to load the catalog from disk, if it fails
+    /// it will build it by querying the Kubernetes API server.
+    /// If querying the API server fails, the function will return an empty catalog.
+    pub async fn new<F, Fut>(resource_catalog_file: PathBuf, build_kubeclient_fn: F) -> Self
```
Comment (fabriziosestito) [inline src/scaffold/admission_request.rs:89]: let's add a comment explaining why we need to defer the creation of the client passing a fn here.

Comment (flavio) [reply src/scaffold/admission_request.rs:93]: done

**Pull request merged**